ouroboros-consensus/src/Ouroboros/Consensus/Ledger/HardFork.hs

```diff
@@ -0,0 +1,433 @@
+
+  type ApplyTxErr (Forked blk1 blk2) = ForkedApplyTxErr blk1 blk2
+
+  applyTx ledgerConfig forkedGenTx forkedLedgerState =
+    case (unForkedGenTx forkedGenTx, unForkedLedgerState forkedLedgerState) of
+      (BeforeFork genTx, BeforeFork ledgerState) ->
+        withExcept BeforeForkApplyTxErr . fmap (ForkedLedgerState . BeforeFork) $
+          applyTx
+            (ledgerConfigBeforeFork ledgerConfig)
+            genTx
+            ledgerState
+      (AfterFork genTx, AfterFork ledgerState) ->
+        withExcept AfterForkApplyTxErr . fmap (ForkedLedgerState . AfterFork) $
+          applyTx
+            (ledgerConfigAfterFork ledgerConfig)
+            genTx
+            ledgerState
+      (BeforeFork _, AfterFork _) -> throwE OldTransactionAfterFork
+      (AfterFork _, BeforeFork _) -> throwE NewTransactionBeforeFork
+
+  reapplyTx ledgerConfig forkedGenTx forkedLedgerState =
+    case (unForkedGenTx forkedGenTx, unForkedLedgerState forkedLedgerState) of
+      (BeforeFork genTx, BeforeFork ledgerState) ->
+        withExcept BeforeForkApplyTxErr . fmap (ForkedLedgerState . BeforeFork) $
+          reapplyTx
+            (ledgerConfigBeforeFork ledgerConfig)
+            genTx
+            ledgerState
+      (AfterFork genTx, AfterFork ledgerState) ->
+        withExcept AfterForkApplyTxErr . fmap (ForkedLedgerState . AfterFork) $
+          reapplyTx
+            (ledgerConfigAfterFork ledgerConfig)
+            genTx
+            ledgerState
+      (BeforeFork _, AfterFork _) -> throwE OldTransactionAfterFork
+      (AfterFork _, BeforeFork _) -> throwE NewTransactionBeforeFork
+
+  reapplyTxSameState ledgerConfig forkedGenTx forkedLedgerState =
+    case (unForkedGenTx forkedGenTx, unForkedLedgerState forkedLedgerState) of
+      (BeforeFork genTx, BeforeFork ledgerState) ->
+        ForkedLedgerState . BeforeFork $
+          reapplyTxSameState
+            (ledgerConfigBeforeFork ledgerConfig)
+            genTx
+            ledgerState
+      (AfterFork genTx, AfterFork ledgerState) ->
+        ForkedLedgerState . AfterFork $
+          reapplyTxSameState
+            (ledgerConfigAfterFork ledgerConfig)
+            genTx
+            ledgerState
+      (BeforeFork _, AfterFork _) ->
+        error "reapplyTxSameState: impossible! OldTransactionAfterFork"
+      (AfterFork _, BeforeFork _) ->
+        error "reapplyTxSameState: impossible! NewTransactionBeforeFork"
+
+deriving instance
+  (Show (GenTx blk1), Show (GenTx blk2)) => Show (GenTx (Forked blk1 blk2))
+
+deriving instance
+  (Eq (GenTxId blk1), Eq (GenTxId blk2)) => Eq (GenTxId (Forked blk1 blk2))
+
+deriving instance
+  (Ord (GenTxId blk1), Ord (GenTxId blk2)) => Ord (GenTxId (Forked blk1 blk2))
+
+
+instance
+  CanHardForkBlock
+    (SimplePBftBlock SimpleMockCrypto PBftMockCrypto)
+    (SimplePraosBlock SimpleMockCrypto PraosMockCrypto) where
+
+  -- Currently hard forking at slot 0 as we cannot fork yet
+  hardForksAt _ = Just $ SlotNo 0
```

Comment:
I was playing around with this number and realised that I forgot to update `hardForksAtView`, which also defines the slot to be 0. I know we need both, but it would be good if they are somehow "linked" together to make sure they are consistent.

Reply:
Hmm, yeah I tried linking them by running `protocolLedgerView` in the one and then using `hardForksAtView`, but that requires a `NodeConfig`, which we don't have at `hardForksAt` call sites. Equally the other way around we can't get a `LedgerState` from the `LedgerView`, so I can't think how to do it generically.

I suppose for a specific case, where you know that the `protocolLedgerView` function doesn't need the `nodeConfig`, then you could do it?

I think in real use cases, these will both be defined as projection functions to the same bit of state, so they'll probably share more in that case.